Playwright: implement ability to log in to WPCOM from a popup.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR implements the ability to log in to WPCOM through a popup window and the last bit of the original `wp-likes` spec.

In addition, some longstanding todo items have also been addressed in this patch:
- migrate `LoginPage` to be an inheritance of `BaseContainer`.
- implement a separate `getAccountCredential` and `getAccountSiteURL`.

#### Testing instructions

As usual:

CI

- ensure tests pass without any issues on TeamCity.

Local

- pull the branch and ensure they run with mocha --config .mocharc_playwright.yml specs/specs-playwright/.

Parent: #53169

worldomonation committed:
Implement new test `Like post as logged out user`.

worldomonation committed:
Change some of the waits, and use the BrowserManager.clearCookies function in the test instead of calling raw commands.

Remove page.waitForNavigation() from LoginPage.login() due to issues with loginFromPopUp() flow.

worldomonation committed:
Migrate LoginPage to use the BaseContainer.

Wait for subtle animation on the LoginPage when a user is already logged in.

Remove space from test names.
Additional code comments and JSDoc on LoginFlow.
Make the Page object parameter in LoginFlow.baseflow an optional param.

worldomonation commented on packages/calypso-e2e/src/browser-manager.ts:
```diff
@@ -147,4 +147,5 @@ export async function close(): Promise< void > {
  */
 export async function clearCookies( page: Page ): Promise< void > {
 	await page.context().clearCookies();
+	await page.waitForTimeout( 250 );
```
I will evaluate if this is actually necessary.

worldomonation commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -34,16 +34,39 @@ export function getCalypsoURL(
  * @throws {Error} If username does not correspond to a valid entry in the configuration file.
  */
-export function getAccountCredential( username: string ): string {
+export function getAccountCredential( username: string ): string[] {
```
I have changed  this method to only return the `username`/`password` as an array of strings.

worldomonation commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -34,16 +34,39 @@ export function getCalypsoURL(
+	try {
+		const [ , , url ] = testAccounts[ username ];
+		if ( url ) {
+			return `https://${ url }`;
```
Playwright needs the full `https://` prefix in order to navigate.

I did not want to rely on an external library like [JS-URI](https://medialize.github.io/URI.js/) just for this, especially since we control the contents of the credentials file.

WunderBart replied:
Would it make sense to use the URL constructor here?

```js
return new URL( `https://${ url }` );
```

I guess the benefit of that would be the validation of the URL string.

worldomonation replied:
> Would it make sense to use the URL constructor here?
> 
> ```js
> return new URL( `https://${ url }` );
> ```
> 
> I guess the benefit of that would be the validation of the URL string.

Yeah I came across another part of my code that used the URL constructor. It would make sense to switch to that for here.

worldomonation commented on packages/calypso-e2e/src/lib/pages/login-page.ts:
```diff
@@ -1,35 +1,40 @@
+	async _postInit(): Promise< void > {
+		await this.page.waitForLoadState( 'networkidle' );
+		const container = await this.page.waitForSelector( selectors.loginContainer );
+		await container.waitForElementState( 'stable' );
```
Wait to handle the subtle fade-in animation if there is already a logged in session.

WunderBart commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -34,16 +34,39 @@ export function getCalypsoURL(
 	} catch ( err ) {
-		throw new Error( `Credential for username: ${ username } not found in configuration file.` );
+		throw new Error(
+			`Configuration file did not contain credentials for requested user ${ username }.`
```
This looks like the same situation as https://github.com/Automattic/wp-calypso/pull/52887#discussion_r635184891. I think this will only throw if the given `username` doesn't exist in the `testAccounts` object.

worldomonation replied:
Good point, though within `local-decypted.json` the test account are keyed like so
```
acountType: [ username, password, site]
```

This function is checking if the `accountType` matches the `username` (which is poor naming by the way).

I propose the parameter variable be renamed to `accountType` to make it clear that this isn't matching on the username, but rather the account classification.

WunderBart commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -34,16 +34,39 @@ export function getCalypsoURL(
+		}
+		throw new Error();
+	} catch ( err ) {
+		throw new Error( `Configuration file did not contain URL for requested user ${ username }.` );
```
Ditto

WunderBart commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -34,16 +34,39 @@ export function getCalypsoURL(
+		if ( url ) {
+			return `https://${ url }`;
+		}
+		throw new Error();
```
Accidental `throw`?

WunderBart replied:
Unless related to https://github.com/Automattic/wp-calypso/pull/53310#discussion_r648139543

worldomonation replied:
Interesting....how did this make it here?

worldomonation replied:
Oh, I remember now.

If `throw new Error()` at Line 67 is not there, then TypeScript transpiler complains about not all paths return a value.

The `throw` there is to ensure that if `url` was undefined, an error is thrown which is caught in the next line.

worldomonation replied:
~~What do you think about the function throwing an error (like it does now) as opposed to say, returning an empty string if `url` is not defined in the secrets file?~~

~~Personally I like the approach of throwing since it would halt execution immediately, since not having a valid `url` to visit is not a problem that the framework should correct.~~

All I said is moot point as `new URL(undefined)` would throw an error on its own, so no need for me to manually throw one.

WunderBart commented on packages/calypso-e2e/src/lib/flows/login-flow.ts:
```diff
@@ -20,37 +20,68 @@ export class LoginFlow {
 	 *
 	 * @returns {Promise<void>} No return value.
 	 */
 	async login(): Promise< void > {
```
nit: Should we make it `logIn` to indicate an action? Probably same goes for the `LogInPage` object.

worldomonation replied:
You're referring to the method name, correct? 

To be honest, I am still struggling applying the Automattic Writing Style Guide, which covers the distinction of Log In and login.

I would like to start another PR for housekeeping and minor cleanups that had been missed in previous code reviews and include this suggestion as one of the task items in that.

worldomonation replied:
I've filed #53556 to address this nit :)

WunderBart commented on packages/calypso-e2e/src/lib/flows/login-flow.ts:
```diff
@@ -20,37 +20,68 @@ export class LoginFlow {
+		await this.baseflow( popupPage );
+
+		// Wait for the popup to be closed before passing control back.
+		await popupPage.waitForEvent( 'close' );
```
These events are really something that's been missing in Selenium ⭐

worldomonation replied:
Yeah, it's really cool that Playwright can listen to events. I'm sure there's a lot more power I am not harnessing, but so far I've been able to do everything I need without going that far.

WunderBart commented on packages/calypso-e2e/src/lib/flows/login-flow.ts:
```diff
@@ -20,37 +20,68 @@ export class LoginFlow {
+	async loginFromPopup(): Promise< void > {
+		// Popup emits the event 'popup'. Capturing the event obtains the Page object
+		// for the popup page, where the login form is located.
+		const popupPage = await this.page.waitForEvent( 'popup' );
```
Because we're expecting the `popup` event here, does it mean that `loginFromPopup` will always have to be called within the `Promise.all` chain to avoid race condition? I.e.

```js
await Promise.all( [
  loginFromPopup(),
  this.page.click( '.login-button' ),
] );
```

worldomonation replied:
Interestingly, yes and no!

The [documentation](https://playwright.dev/docs/multi-pages/#handling-popups) covers two cases: when the action that triggers the popup is known, and unknown.

If it's known then one would use the `Promise.all` chain.

My implementation here is the second case, where I treat the action that causes the popup to be 'unknown'. In that case, Playwright can still listen to the `popup` event emitted and that would grab the page handle referencing the popup page. Pretty neat.

I chose this approach because of a couple of reasons:
- generic: I wanted the `LoginFlow.loginFromPopup()` to be able to stand on its own, for it to handle all login steps on its own.
- loose coupling: I did not want the pages or tests to be tasked with the process of triggering the popup, wait for the page, then pass the page to LoginFlow for executing the action.

As always, with that said, I am open to changing the approach here! So far this approach seems to work just fine, but if I run into issues later on I'd be open to delegating the popup handling to the page/specs.

WunderBart replied:
>My implementation here is the second case, where I treat the action that causes the popup to be 'unknown'. In that case, Playwright can still listen to the popup event emitted and that would grab the page handle referencing the popup page. Pretty neat.

Thanks for the explanation! This makes that approach totally valid, although still needs specific handling either way. Maybe we could just link to the `popups` doc to make folks aware about a proper usage of this method (known vs unknown trigger)?

worldomonation replied:
Added link to the documentation.

WunderBart commented on packages/calypso-e2e/src/lib/pages/published/published-posts-list-page.ts:
```diff
@@ -24,6 +24,15 @@ export class PublishedPostsListPage extends BaseContainer {
+	 * @returns {Promise<void>} No return value.
+	 */
+	async _postInit(): Promise< void > {
+		await this.page.waitForSelector( selectors.posts );
```
I can see this pattern repeating, where we wait for additional elements in the `_postInit` hook. Do you think making the `BaseContainer` accept an array of selectors that need to be awaited before initialization would help here? The `_init` method could then be:

```js
async _init(): Promise< void > {
	await this._visit();
	await this._preInit();
	await this.page.waitForLoadState( 'load' );
	if ( this.selectors.length > 0 ) {
		await Promise.all(
			this.selectors.map( selector => this.page.waitForSelector( selector ) )
		);
	}
	await this._postInit();
}
```

worldomonation replied:
You are right, I do frequently use the `_postInit` more or less to wait on additional elements. 

I believe the Selenium counterpart accepts a list of elements. I'll spin up a new task on the board to alter this behavior. How does that aound?

worldomonation replied:
Filed #53557 for the task.

worldomonation committed:
Address code review comments.

- clarify the role of username in getAccount* functions of DataHelper.
- remove wait from BrowserManager.clearCookies.

worldomonation committed:
Improve error handling in DataHelper.getAccountSiteURL.

Need the last comma in the destructuring in order to remove the URL portion of the returned value.

Errant console.log.
Change how the getAccountCredential and getAccountSiteURL funtions handle data retrieval.

Added comment as requested in PR.

WunderBart commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -32,19 +32,44 @@ export function getCalypsoURL(
+
+	const [ , , url ] = testAccounts[ accountType ];
+	if ( ! url ) {
+		throw new TypeError( `Secrets entry for ${ accountType } has no site URL defined.` );
```
nit: this looks more like a [`ReferenceError`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ReferenceError)
```suggestion
		throw new ReferenceError( `Secrets entry for ${ accountType } has no site URL defined.` );
```

WunderBart reviewed (approved):
Together with https://github.com/Automattic/wp-calypso/pull/53291 looks good! Just needs a rebase. Thanks! 🙌

worldomonation committed:
Rebase and fix.

worldomonation committed:
Rebase hell

worldomonation committed:
Change type of error.

worldomonation commented:
> Together with #53291 looks good! Just needs a rebase. Thanks! 🙌

Thanks for the review as always! This PR ended up being quite the rebase hell 😱

worldomonation committed:
Rebase artifacts again, this time the two `wp-likes` tests.

worldomonation committed:
Address eslint issues.
